### PR TITLE
feat: Add aggregate parameters to vega-transform, and exponential moving average

### PR DESCRIPTION
**Pull request by Xitian9**
This adds exponential moving averages, fixing #3341. In order to do this, it enables supplying parameters to aggregation functions. This addresses the block on https://github.com/vega/vega-lite/issues/4439.

As I'm not a javascript programmer, feedback will be gratefully taken on board.

### CODE REVIEW

**Comment by Xitian9** [comment]
Is there any interest in this?

**Comment by domoritz** [comment]
Thanks for the pull request. I just approved the test run again.

**Comment by Xitian9** [comment]
If/when this is merged, #3577 could be made more general by passing the requested quantiles as a parameter.

**Comment by shcheklein** [comment]
Hey @Xitian9 thanks for the effort! This is exactly what we are looking for in this case https://github.com/iterative/vscode-dvc/issues/3837 . In ML tracker tools it's very common to use the exponential moving average. I'm still looking into this (and learning along the way if there are other ways to achieve this).

@domoritz are there any blockers left to merge this? (maybe we can also join and help). Also, if this is merged will it become available automatically in vega-lite?

File: packages/vega-transforms/src/Aggregate.js
Comment (domoritz) [inline packages/vega-transforms/src/Aggregate.js:14]: Do we need to repeat the name `aggregate` here or is it already obvious because of the context of the aggregate?

**Review by domoritz — APPROVED** [review]
Looks reasonable to me. The API question I have is whether it makes sense for the parameter to be a separate property or somehow part of the ops. I guess this is consistent with the rest of the API. 

@jheer needs to make a final pass and merge

**Comment by shcheklein** [comment]
Okay, was looking into this one more time and I don't get so far is why this is considered the _moving_ average. It's producing a single value. @Xitian9 could you clarify please the use case for this? can it be used to produce a moving average, or may be I'm missing something.

File: packages/vega-transforms/src/Aggregate.js
Comment (Xitian9) [reply packages/vega-transforms/src/Aggregate.js:14]: I was going by the naming agreed upon in https://github.com/vega/vega-lite/issues/4439. We cannot just call it params.params, as this would conflict with the params argument in windowing functions. Anytime you would want to apply an aggregate with parameter to over a window with a parameter, you would run into problems.

**Comment by Xitian9** [comment]
> Can it be used to smooth a linear plot or does it calculate a single number?

The aggregate in itself produces a single number (which is why I call it `exponential` instead of `exponentialmovingaverage` or something like that). However, it can easily be combined with a windowing function to produce a smoothing, which is the use case I have in mind as well.

> @domoritz are there any blockers left to merge this? (maybe we can also join and help). Also, if this is merged will it become available automatically in vega-lite?

I don't believe it would automatically be available in vega-lite, and that would need to be updated as well.

**Comment by shcheklein** [comment]
@jheer sorry for bothering you, anything we can do (help with) here to unblock this and move forward? Please let me know.

**Comment by Xitian9** [comment]
Is there any appetite for this? If not I'll delete the PR.

**Comment by shcheklein** [comment]
@Xitian9 please keep it for a while :) We definitely need this! I can't make decisions here and / or review, but please keep it possible. I hope folks will get some time to merge it. Thanks for your work.

**Comment by shcheklein** [comment]
Hey folks, @jheer , any chance we can move it forward? Could we at least identify what is blocking this PR atm?

**Comment by jheer** [comment]
Thanks for the PR, sorry for the delay in reviewing!

I checked with @jonmmease and expanding aggregates this way does not have negative downstream consequences for VegaFusion (other than that weighted exponential aggregates will not be optimized by VegaFusion at the moment).

**Comment by julieg18** [comment]
Hello, I'm working on adding `aggregate_parameters` into vega-lite (https://github.com/vega/vega-lite/issues/9200), but when I tested using the params in a vega editor spec, all I could get was `NaN`:

image

I tried a couple different field variations but `NaN` was always the result. Have I misunderstood what `aggregate_params` is supposed to consist of? Here's the [example vega editor spec](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhAATJJhSoA2qHFIEcNCAhIANnAgB9OAA8NOA9JDF9DQ2lUgyAJzYMcaAAwzdBiGgALJQATF4CMjhILhrelF4JAJwAvlKgru6eqD46+g6owYFeIQBskdGx2fGBiSWp6W4eaACMvnkBqEIhza0gUTEIccUlABz1zo1ZvX75wQDsIwIh5QNDI93jGU2oy7n+aEJLgQDMK5Ve1XMhm5NouzMdx5QlC7v95-FzzQI3md5t+wKlBGIzKfQqgyqXkScxSaQmf1Q03aB2eIUSIzOkIuXm+cIaiORgKExxhgSxQ0CIWu8K2WXuKNQTzmpMSFKhiVJyQAuuN1JptO8EEZbHp7NZRfYOhcfvD+Vp0GYcBI4OJMDR9EYkMQyNYxAwXFAFXtDCZzAhLFoZJgYuIIAAzNguSFOTAATxwxqQZFcfEwVpA9pocD0sg6yhNAW5MjYeEcICVKrVGr0IGjICQ4ZALkMDD0mDTMjpTDd8bphYzPpzFH9RiFWaFIrscEo8kUaZ5naAA) I was using.

File: packages/vega-transforms/src/Aggregate.js
Comment (domoritz) [reply packages/vega-transforms/src/Aggregate.js:49]: @Xitian9

File: packages/vega-transforms/src/util/AggregateOps.js
Comment (domoritz) [inline packages/vega-transforms/src/util/AggregateOps.js:138]: Isn't an init missing here? Otherwise `m.exp_r` isn't assigned.

**Comment by domoritz** [comment]
@julieg18 hmm, yeah, something is wrong here. The types look wrong and the implementation may also be missing an init.

I think the aggregate parameter is just a value, not a field judging from the code but maybe I'm missing something: [Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhAATJJhSoA2qHFIEcNCAhIANnAgB9OAA8NOA9JDF9DQ2lU79D1ABZKAJgAMAgL5SoLoGEGgebt6eAGwBQS6hqEKeAIzJsc4hYZQA7AAcAp7pwa5CBW4AzEXxaOWUUXmFgRmuHrm5MU3FCUJRngCcuVWZiZTlfdluQ6612WN9fgC66eqa2mY4EnDimDT6RkjEZNZiDABOUFroXSbmCJZaMpinSOIQAGZspwiOoJgAnjhLiAkGQyKc+JgHiA3jQ4HpZAllM1QgsZGw8I4QOtNttdnoQKjgYiQOCIAw9JgCTIQWCIXAjDgkM8EIjvJQBEtOX4gA) (still fails because of some error somewhere). 

@Xitian9 @shcheklein can you post an example you used to test this feature?

File: packages/vega-transforms/src/Window.js
Comment (domoritz) [inline packages/vega-transforms/src/Window.js:41]: Is the default correct here? There may be more than two ops.

File: packages/vega-transforms/src/Aggregate.js
Comment (domoritz) [inline packages/vega-transforms/src/Aggregate.js:14]: Shouldn't this be `[params.aggregate_params]` since we cannot assume a single op?

File: packages/vega-transforms/src/Window.js
Comment (Xitian9) [reply packages/vega-transforms/src/Window.js:41]: The code will pad aggregate_params with nulls as necessary to match the number of ops.

**Comment by Xitian9** [comment]
> @Xitian9 @shcheklein can you post an example you used to test this feature? Also, could you send a follow up pull request to add docs.

Test code is here: https://github.com/vega/vega/blob/main/packages/vega-transforms/test/aggregate-test.js

File: packages/vega-transforms/src/Aggregate.js
Comment (Xitian9) [reply packages/vega-transforms/src/Aggregate.js:14]: The code will pad aggregate_params with nulls as necessary to match the number of ops.

**Comment by Xitian9** [comment]
> Hello, I'm working on adding `aggregate_parameters` into vega-lite ([vega/vega-lite#9200](https://github.com/vega/vega-lite/issues/9200)), but when I tested using the params in a vega editor spec, all I could get was `NaN`:
> image
> 
> I tried a couple different field variations but `NaN` was always the result. Have I misunderstood what `aggregate_params` is supposed to consist of? Here's the [example vega editor spec](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhAATJJhSoA2qHFIEcNCAhIANnAgB9OAA8NOA9JDF9DQ2lUgyAJzYMcaAAwzdBiGgALJQATF4CMjhILhrelF4JAJwAvlKgru6eqD46+g6owYFeIQBskdGx2fGBiSWp6W4eaACMvnkBqEIhza0gUTEIccUlABz1zo1ZvX75wQDsIwIh5QNDI93jGU2oy7n+aEJLgQDMK5Ve1XMhm5NouzMdx5QlC7v95-FzzQI3md5t+wKlBGIzKfQqgyqXkScxSaQmf1Q03aB2eIUSIzOkIuXm+cIaiORgKExxhgSxQ0CIWu8K2WXuKNQTzmpMSFKhiVJyQAuuN1JptO8EEZbHp7NZRfYOhcfvD+Vp0GYcBI4OJMDR9EYkMQyNYxAwXFAFXtDCZzAhLFoZJgYuIIAAzNguSFOTAATxwxqQZFcfEwVpA9pocD0sg6yhNAW5MjYeEcICVKrVGr0IGjICQ4ZALkMDD0mDTMjpTDd8bphYzPpzFH9RiFWaFIrscEo8kUaZ5naAA) I was using.

"exponential" takes a number as an argument, and you have provided the string "param_value". Do you still get NaN if you supply a double?

**Comment by Xitian9** [comment]
> @julieg18 hmm, yeah, something is wrong here. The types look wrong to me.
> 
> I think the aggregate parameter is just a value, not a field judging from the code but maybe I'm missing something:

I'm not a javascript programmer, so I could definitely believe I got the types wrong. The type needs to be generic enough to accept any sort of value, as different ops will have different argument types. For "exponential" that's a number, but for other as-yet-unimplemented ops it could be something of arbitrary complexity.

**Comment by julieg18** [comment]
> "exponential" takes a number as an argument, and you have provided the string "param_value". 

Makes sense, thanks! As previously mentioned, I can only provide strings since currently the type is set to fields. 

> Do you still get NaN if you supply a double?

Apologies, I'm not sure what you mean by double? Do you mean something like this:

image

I would still get `NaN` in [this case](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhAATJJhSoA2qHFIEcNCAhIANnAgB9OAA8NOA9JDF9DQ2lUgyAJzYMcaAAwzdBiGgALJQATF4CMjhILhrelF4JAJwAvlKgru6eqD46+g6owYFeIQBskdGx2fGBiSWp6W4eaACMvnkBqEIhza0gUTEIccUlABz1zo1ZvX75wQDsIwIh5QNDI93jGU2oy7n+aEJLgQDMK5Ve1XMhm5NouzMdx5QlC7v95-FzzQI3md5t+wKlBGIzKfQqgyqXkScxSaQmf1Q03aB2eIUSIzOkIuXm+cIaiORgKExxhgSxQ0CIWu8K2WXuKNQTzmpMSFKhiVJyQAuuN1JptGYcBI4OJMDR9EYkMQyNYxAwXFAtOgHiZzAhLFoZJgYuIIAAzNguSFOTAATxwypASDIrj4mC1IH1NDgelkHWU4NWAIc3JkbDwjh0DEGMiFIrFEr0ID91o9Xo0Rlsens1hchgYekwMZkNrtFAdRneCA94kzeixSbsWl5teSQA)

**Comment by domoritz** [comment]
> Makes sense, thanks! As previously mentioned, I can only provide strings since currently the type is set to fields.

The editor will only warn you. Numbers are still passed to Vega.

**Comment by julieg18** [comment]
> The editor will only warn you. Numbers are still passed to Vega.

Perhaps I'm misunderstanding something, but [inputting numbers causes errors to be thrown](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhAATJJhSoA2qHFIEcNCAhIANnAgB9OAA8NOA9JDF9DQ2lUgyAJzYMcaAAwzdBiGgALJQATF4CMjhILhrelF4JAJwAvlKgru6eqD46+g6owYFeIQBskdGx2fGBiSWp6W4eaACMvnkBqEIhza0gUTEIccUlABz1zo1ZvX75wQDsIwIh5QNDI93jGU2oy7n+aEJLgQDMK5Ve1XMhm5NouzMdx5QlC7v95-FzzQI3md5t+wKlBGIzKfQqgyqXkScxSaQmf1Q03aB2eIUSIzOkIuXm+cIaiORgKExxhgSxQ0CIWu8K2WXuKNQTzmpMSFKhiVJyQAuuN1JptO8EEZbHp7NZRfYOhcfvD+Vp0GYcBI4OJMDR9EYkMQyNYxAwXFAFXtDCZzAhLFoZJgYuIIAAzNguSFOTAATxwxqQZFcfEwVpA9pocD0sg6yhNAW5MjYeEcICVKrVGr0IGjICQ4ZALkMDD0mDTMjpTDd8bphYzPpzFH9RiF4ZlvKbySAA) and nothing get rendered, in the view or data viewer: 

image

**Comment by julieg18** [comment]
> Types are now fixed thanks to https://github.com/vega/vega/pull/3846, but as expected, we are still hitting the Unexpected token ']' error when trying to [run a spec](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhAATJJhSoA2qHFIEcNCAhIANnAgB9OAA8NOA9JDF9DQ2lU79D1ABZKAJgAMAgL5SoLoGEGgebt6eAGwBQS6hqEKeAIzJsc4hYZQA7AAcAp7pwa5CBW4AzEXxaOWUUXmFgRmuHrm5MU3FCUJRngCcuVWZiZTlfdluQ6612WN9fgC66eqa2mY4EnDimDT6RkjEZNZiDABOUFroXSbmCJZaMpinSOIQAGZspwiOoJgAnjhLiAkGQyKc+JgHiA3jQ4HpZAllM1QgsZGw8I4QOtNttdnoQKjgYiQOCIAw9JgCTIQWCIXAjDgkM8EIjvJQBEtOX4gA).

Actually, I tested `aggregate_params` today with the latest version of vega in my local vega editor and everything seems to be working! 

image

Not sure why I was getting the error previously, but unless it shows up again, I'll get back to https://github.com/vega/vega-lite/issues/9200

**Comment by domoritz** [comment]
Awesome. Thanks for the update. I'm glad it's working now.

**Pull request merged**
